feat(aws-fsx): construct for NetApp ONTAP file system

badmintoncryer opened the pull request:
### Issue # (if applicable)

Closes #41.

### Reason for this change

There is no L2 construct for creating FSx for NetApp ONTAP file system.

### Description of changes

- Add `OntapFileSystem` class
- Add `MaintenanceTime` class
  - However `LusterMaintenanceTime` class already exists in `aws-cdk-lib.aws-fsx` module , it seems not to be used in other file systems such as NetApp ONTAP. Therefore, I newly added `MaintenanceTime` class.

```ts
new ocf.aws_fsx.OntapFileSystem(stack, 'OntapMultiAzFileSystem', {
  vpc,
  vpcSubnets: vpc.privateSubnets,
  storageCapacityGiB: 5120,
  ontapConfiguration: {
    automaticBackupRetention: Duration.days(7),
    dailyAutomaticBackupStartTime: new aws_fsx.DailyAutomaticBackupStartTime({
      hour: 1,
      minute: 0,
    }),
    deploymentType: ocf.aws_fsx.OntapDeploymentType.MULTI_AZ_2,
    diskIops: 15360,
    endpointIpAddressRange: '192.168.39.0/24',
    fsxAdminPassword: 'fsxPassword1',
    haPairs: 1,
    preferredSubnet: vpc.privateSubnets[0],
    routeTables: [vpc.privateSubnets[0].routeTable, vpc.privateSubnets[1].routeTable],
    throughputCapacity: 384,
    weeklyMaintenanceStartTime: new ocf.aws_fsx.MaintenanceTime({
      day: aws_fsx.Weekday.SUNDAY,
      hour: 1,
      minute: 0,
    }),
  },
  removalPolicy: RemovalPolicy.DESTROY,
});
```

The CI job failed due to a conflict with the peer dependency of `aws-cdk-lib`.
I would like to set the peer dependency version to the newest one.

https://github.com/open-constructs/aws-cdk-library/actions/runs/10023311572/job/27704071917?pr=42

### Description of how you validated changes

Added both unit and integ tests.

### Checklist

- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/open-constructs/aws-cdk-library/blob/main/CONTRIBUTING.md)

---
_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
Could you check if they are the `Token`?

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
Could you check if they are the Token?

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
I want to pay special attention to the divisor.

Although the `validateHaPairs` method checks that `haPair` is greater than or equal to 1, this is a separate method and the situation could easily change. If other developers change the order of the validation methods in the future or make changes to the way the `haPair` is checked in the `validateHaPairs` method, can we guarantee that there will always be no assignment of 0? If not, the CDK will panic in this area.

In TypeScript, division by 0 is not an error, but when converted to other languages by jsii, there must be languages where it is an error (e.g. Golang).

Since this project specifically wants to guarantee safety and reliability, how about checking that the `haPair` is greater than 0 in this part as well? Although this case may not be necessary in most cases, it is better to be vigilant.

What do you think? Even if you think there is no need for this at all, at least adding a comment (that there is no possibility of 0 being passed on) to the code would be sufficient precaution. (There may be many developers like me who are sensitive to division by zero ...)

If you know of a better way, let me know.

badmintoncryer replied:
I'm completely agree with your opinion! I've added below code to validate zero division.

```ts
    if (haPair <= 0 || !Number.isInteger(haPair)) {
      throw new Error(`'haPair' must be a positive integer, got ${haPair}`);
    }
```

go-to-k commented on src/aws-fsx/maintenance-time.ts:
Is this class for use outside of ONTAP? Or is it ONTAP specific?

go-to-k reviewed (changes requested):
Thanks for your great contribution!

The amount of changes is too large to look at all at once, so I have commented on the few superficial things that need to be fixed. I would like to look at the core later when I have more time, but it may take some time to look at everything.

If you have links to detailed documentation that we reviewers should have seen beforehand, or your design or implementation intentions, it would be helpful if you could include them in the PR description or code comments. This would improve the quality and speed of my review and may make it easier for other reviewers to participate in the review.

go-to-k commented on src/aws-fsx/maintenance-time.ts:
```diff
@@ -0,0 +1,74 @@
+  readonly minute: number;
+}
+
+/**
+ * Class for scheduling a weekly maintenance time.
+ */
+export class MaintenanceTime {
```
Is this class for use outside of ONTAP? Or is it ONTAP specific?

badmintoncryer replied:
I think this class is used for more than just ONTAP.

There is a LustreMaintenanceTime class in aws-cdk-lib that is similar to the MaintenanceTime class. I've added the MaintenanceTime class because the LustreMaintenanceTime class appears to be specific to the Lustre file system. By using the MaintenanceTime class, we can apply it to other FSx file system constructs as well.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,498 @@
+   *
+   * You can have overlapping endpoint IP addresses for file systems deployed in the same VPC/route tables, as long as they don't overlap with any subnet.
+   *
+   * @default -
```
I've added!

I referenced [this document](https://docs.aws.amazon.com/fsx/latest/APIReference/API_CreateFileSystemOntapConfiguration.html#endpoint-ip-address-range).

```
By default in the Amazon FSx API, Amazon FSx selects an unused IP address range for you from the 198.19.* range.
```

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,498 @@
+   *
+   * This value is required when DeploymentType is set to MULTI_AZ_1 or MULTI_AZ_2.
+   *
+   * @default
```
Here’s how I wrote it. What do you think?

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,498 @@
+   */
+  readonly endpointIpAddressRange?: string;
+
+  /**
+   * The ONTAP administrative password for the fsxadmin user with which you administer your file system using the NetApp ONTAP CLI and REST API.
+   */
+  readonly fsxAdminPassword?: string;
```
I couldn't find any description, so I'll try to create file system with not defining this parameter.

badmintoncryer replied:
I tried deploying with and without specifying the `fsxAdminPassword `, but unfortunately, I couldn't verify the password through the GUI, so the default behavior remains unclear. 
As a countermeasure, I considered making the password parameter mandatory. What do you think?

go-to-k replied:
The documentation says that with quick start, the password will not be set. So I assume this is also the case in this case, what do you think?

> If you created the file system using the Quick create option, the fsxadmin password was not set, so you'll to set one in order to log in to the ONTAP CLI.

https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/managing-resources-ontap-apps.html#fsxadmin-ontap-cli

Also, if it is unspecified, it appears that the user cannot log in. (Might be better to actually try it...).
In other words, it is secure, and users would be reassured if this is also mentioned in the JSDoc.

Perhaps some people do not want to set a password in IaC or want to set it later in the console. If it were made mandatory, those people would be in trouble.

badmintoncryer replied:
Than you very much for your great suggestion!! 
I updated the JSDoc based on the content you provided.

badmintoncryer commented:
@go-to-k Thank you for your review! I've addressed most of your comments.(Some of them still haven't updated.. I'll do it later)
I'll update my PR description to add link of documentation or something. Please wait for a while.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,514 @@
+      return;
+    }
+    if (
+      automaticBackupRetention.toMilliseconds() < Duration.days(1).toMilliseconds() ||
```
This validation is based on these description.

> AutomaticBackupRetentionDays
> The number of days to retain automatic backups. Setting this property to 0 disables automatic backups. You can retain automatic backups for a maximum of 90 days. The default is 30.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-automaticbackupretentiondays

badmintoncryer replied:
I've added that!

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,514 @@
+
+    const automaticBackupDisabled = !automaticBackupRetention || automaticBackupRetention?.toDays() === 0;
+
+    if (dailyAutomaticBackupStartTime && automaticBackupDisabled) {
```
The default value of `automaticBackupRetention` is 0 in this construct. This means disable automatic backup.

```ts
automaticBackupRetentionDays: ontapConfiguration.automaticBackupRetention?.toDays() ?? 0,
```
https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R278

Because `dailyAutomaticBackupStartTime` can be set for only automatic backup is enabled, I added this validation.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
This validation is based on these description.

> The minimum and maximum values for this property depend on the value of HAPairs and StorageCapacity. The minimum value is calculated as StorageCapacity * 3 * HAPairs (3 IOPS per GB of StorageCapacity). The maximum value is calculated as 200,000 * HAPairs.


https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-diskiopsconfiguration.html#cfn-fsx-filesystem-diskiopsconfiguration-iops

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
`endpointIpAddressRange` is used for only multi-AZ file system.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-endpointipaddressrange

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
This regex patter is described in cfn docs.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-endpointipaddressrange

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
This pattern is described in cfn docs.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-fsxadminpassword

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
These rules are not documented but I failed to deploy in these condition. Therefore, I added this validation.

https://dev.classmethod.jp/articles/fsxn-password-cannot-be-used-within-the-last-six-times/

badmintoncryer replied:
I hadn't realized that... Cloudformation document says 50...

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-fsxadminpassword

Which value should be accept?

go-to-k replied:
This would no longer be necessary: https://github.com/open-constructs/aws-cdk-library/pull/42#discussion_r1781414058

badmintoncryer commented:
I'll add comments later.

mbonig commented on .projenrc.ts:
```diff
@@ -1,7 +1,7 @@
 import { ArrowParens, NodePackageManager } from 'projen/lib/javascript';
 
-let cdkVersion = '2.120.0';
+let cdkVersion = '2.149.0';
```
we should try to keep this as low as possible. is there a need for this being 2.149.0 vs 2.120.0?

badmintoncryer replied:
I wan to use `DailyAutomaticBackupStartTime` class implemented in [v2.148.0](https://github.com/aws/aws-cdk/releases/tag/v2.148.0)

https://github.com/aws/aws-cdk/pull/30343

It is also possible to address this by describing the same content as `DailyAutomaticBackupStartTime` class in this PR while keeping it at version v2.120.0. Do you have any thoughts on this?

hoegertn replied:
We only want to bump CDK versions if really necessary and in a coordinated way. So CDK version bumps would be its own PR; then constructs can start using it

badmintoncryer replied:
Sure. I've added `daily-automatic-backup-start-time.ts`.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
Multiple HA pairs are only supported for `SINGLE_AZ_2` deployment type.

https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/adding-HA-pairs.html

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
`preferredSubnet` must be specified for Multi-AZ file system.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-preferredsubnetid

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
These values are described in cfn docs. 

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-fsx-filesystem-ontapconfiguration.html#cfn-fsx-filesystem-ontapconfiguration-throughputcapacityperhapair

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
This validation is based on this [description](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fsx-filesystem.html#cfn-fsx-filesystem-storagecapacity).

go-to-k commented:
Hi, @badmintoncryer 

Thanks for the references! It's very helpful!

How are the revisions coming along? I will resume the review when you are ready. (I'm aware that you are still in the process of correcting the following and other points.)

https://github.com/open-constructs/aws-cdk-library/pull/42#discussion_r1702504862

badmintoncryer commented:
@go-to-k I'm really sorry for forgetting this issue😓 I'll resolve it and get back to you later.

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+  /**
+   * The ONTAP administrative password for the `fsxadmin` user with which you administer your file system using the NetApp ONTAP CLI and REST API.
+   *
+   * If you don't specify a password, Amazon FSx will not set one.
```
How about this addition to reassure users?

```suggestion
   * If you don't specify a password, Amazon FSx will not set one. In that case, the user will not be able to log in.
```

go-to-k commented:
Could you resolve the conflicts?

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/administering-file-systems.html#HA-pairs
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/supported-fsx-clients.html#using-block-storage
+   *
+   * @default - 1 HA pair
```
`1 HA pair` is different from 1 ? In other words, why not `@default 1`?

badmintoncryer replied:
There's no deep meaning! I've fixed it to 1.

go-to-k replied:
Really? Did you push your commit?

go-to-k replied:
I could confirm it!

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+  readonly fsxAdminPassword?: string;
+
+  /**
+   * How many high-availability (HA) pairs of file servers will power your file system.
+   * First-generation file systems are powered by 1 HA pair.
```
```suggestion
   * How many high-availability (HA) pairs of file servers will power your file system.
   * 
   * First-generation file systems are powered by 1 HA pair.
```

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   * Second-generation multi-AZ file systems are powered by 1 HA pair.
+   * Second generation single-AZ file systems are powered by up to 12 HA pairs.
+   *
+   * The value of this property affects the values of StorageCapacity, Iops, and ThroughputCapacity.
```
If these are referring to props properties, it is easier to read if they are enclosed in back quotation marks.

```suggestion
   * The value of this property affects the values of `storageCapacity`, `iops`, and `throughputCapacity`.
```

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+  /**
+   * The subnet in which you want the preferred file server to be located.
+   *
+   * This value is required when DeploymentType is set to MULTI_AZ_1 or MULTI_AZ_2.
```
```suggestion
   * This value is required when `deploymentType` is set to `MULTI_AZ_1` or `MULTI_AZ_2`.
```

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   */
+  readonly preferredSubnet?: aws_ec2.ISubnet;
+
+  /**
+   * The route tables in which Amazon FSx creates the rules for routing traffic to the correct file server.
+   * You should specify all virtual private cloud (VPC) route tables associated with the subnets in which your clients are located.
```
```suggestion
  /**
   * The route tables in which Amazon FSx creates the rules for routing traffic to the correct file server.
   *
   * You should specify all virtual private cloud (VPC) route tables associated with the subnets in which your clients are located.
```

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   *
+   * You can define either the `throughputCapacityPerHaPair` or the `throughputCapacity` when creating a file system, but not both.
+   *
+   * For SINGLE_AZ_1 and MULTI_AZ_1 file systems, valid values are 128, 256, 512, 1024, 2048, or 4096 MBps.
+   * For SINGLE_AZ_2, valid values are 1536, 3072, or 6144 MBps.
+   * For MULTI_AZ_2, valid values are 384, 768, 1536, 3072, or 6144 MBps.
+   *
+   * @default - recommended throughput capacity based on the storage capacity
+   */
+  readonly throughputCapacityPerHaPair?: number;
```
Could it be expressed using the Enum-Like class? This is a very complex and exclusive option, so I think it would be a match. (This is just an idea, I could be wrong though.)

If this is possible, it would be even cooler to implement if the validation rules could be confined within the class. (I don't think it to work that easily, but...)

badmintoncryer replied:
I've implemented `ThroughputCapacityPerHaPair` class.

go-to-k replied:
I have also implemented it. I think this is safer and validation can be closed within the class. What do you think?

```ts
export abstract class ThroughputCapacityPerHaPair {
  protected static calculateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number): number {
    if (haPairs <= 0 || !Number.isInteger(haPairs)) {
      throw new Error(`'haPairs' must be a positive integer, got ${haPairs}`);
    }
    return throughputCapacity / haPairs;
  }

  public abstract readonly allowedCapacity: number[];
  public abstract readonly deploymentType: OntapDeploymentType;

  protected constructor(public readonly capacity: number) {}
  protected validateCapacity() {
    if (!this.allowedCapacity.includes(this.capacity)) {
      throw new Error(
        `'throughputCapacity / haPairs' must be one of the following values for ${
          this.deploymentType
        }: ${this.allowedCapacity.join(', ')}. got: ${this.capacity} MB/s/HA pair`,
      );
    }
  }
}

export class SingleAz1ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_128 = new SingleAz1ThroughputCapacityPerHaPair(128);
  public static readonly MB_PER_SEC_256 = new SingleAz1ThroughputCapacityPerHaPair(256);
  public static readonly MB_PER_SEC_512 = new SingleAz1ThroughputCapacityPerHaPair(512);
  public static readonly MB_PER_SEC_1024 = new SingleAz1ThroughputCapacityPerHaPair(1024);
  public static readonly MB_PER_SEC_2048 = new SingleAz1ThroughputCapacityPerHaPair(2048);
  public static readonly MB_PER_SEC_4096 = new SingleAz1ThroughputCapacityPerHaPair(4096);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new SingleAz1ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();
  }

  public readonly allowedCapacity = [128, 256, 512, 1024, 2048, 4096];
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_1;
}

export class MultiAz1ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_128 = new MultiAz1ThroughputCapacityPerHaPair(128);
  public static readonly MB_PER_SEC_256 = new MultiAz1ThroughputCapacityPerHaPair(256);
  public static readonly MB_PER_SEC_512 = new MultiAz1ThroughputCapacityPerHaPair(512);
  public static readonly MB_PER_SEC_1024 = new MultiAz1ThroughputCapacityPerHaPair(1024);
  public static readonly MB_PER_SEC_2048 = new MultiAz1ThroughputCapacityPerHaPair(2048);
  public static readonly MB_PER_SEC_4096 = new MultiAz1ThroughputCapacityPerHaPair(4096);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new MultiAz1ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [128, 256, 512, 1024, 2048, 4096];
  public readonly deploymentType = OntapDeploymentType.MULTI_AZ_1;
}

export class SingleAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_1536 = new SingleAz2ThroughputCapacityPerHaPair(1536);
  public static readonly MB_PER_SEC_3072 = new SingleAz2ThroughputCapacityPerHaPair(3072);
  public static readonly MB_PER_SEC_6144 = new SingleAz2ThroughputCapacityPerHaPair(6144);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new SingleAz2ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [1536, 3072, 6144];
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_2;
}

export class MultiAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  public static readonly MB_PER_SEC_384 = new MultiAz2ThroughputCapacityPerHaPair(384);
  public static readonly MB_PER_SEC_768 = new MultiAz2ThroughputCapacityPerHaPair(768);
  public static readonly MB_PER_SEC_1536 = new MultiAz2ThroughputCapacityPerHaPair(1536);
  public static readonly MB_PER_SEC_3072 = new MultiAz2ThroughputCapacityPerHaPair(3072);
  public static readonly MB_PER_SEC_6144 = new MultiAz2ThroughputCapacityPerHaPair(6144);

  public static validateThroughputCapacityPerHaPair(throughputCapacity: number, haPairs: number) {
    const capacity = ThroughputCapacityPerHaPair.calculateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    const throughputCapacityPerHaPair = new MultiAz2ThroughputCapacityPerHaPair(capacity);
    throughputCapacityPerHaPair.validateCapacity();

  }

  public readonly allowedCapacity = [384, 768, 1536, 3072, 6144];
  public readonly deploymentType = OntapDeploymentType.MULTI_AZ_2;
}

// ...
// ...

private validateThroughputCapacity(
  throughputCapacity?: number,
  throughputCapacityPerHaPair?: ThroughputCapacityPerHaPair,
  haPair: number = 1,
): void {
  // ...
  // ...
  if (throughputCapacityPerHaPair.deploymentType !== this.deploymentType) {
    throw new Error('Error msg')
  }
  // ...
  // ...
  switch (this.deploymentType) {
    case OntapDeploymentType.SINGLE_AZ_1:
      SingleAz1ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.MULTI_AZ_1:
      MultiAz1ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.SINGLE_AZ_2:
      SingleAz2ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
    case OntapDeploymentType.MULTI_AZ_2:
      MultiAz2ThroughputCapacityPerHaPair.validateThroughputCapacityPerHaPair(throughputCapacity, haPairs);
  }
}
```

badmintoncryer replied:
Thank a lot!! It is really nice to implement `{DeploymentType}ThroughputCapacityPerHaPair` for each. I'll update my code like this later.

One more point, I’m considering removing the `throughputCapacity` argument. The reason is that while proper validation is applied for `throughputCapacityPerHaPair`, `throughputCapacity` allows any number, which seems like a missed opportunity.

We could define a similar class for `throughputCapacity`, but I believe using just `throughputCapacityPerHaPair` would be sufficient. In fact, having both options might cause confusion for users. What are your thoughts on this?

go-to-k replied:
> One more point, I’m considering removing the throughputCapacity argument.

Where are you talking about? (Is it `validateThroughputCapacityPerHaPair` method?)

Is it that the validation for `throughputCapacity` and `haPairs` is not necessary? If the value of `throughputCapacity / haPairs` is not included in allowed capacity patterns, doesn't CFn fail?

badmintoncryer replied:
Sorry for the confusion. What I was referring to was not the validate function, but rather the idea of removing `throughputCapacity` from the `OntapConfiguration`.

It was quite unclear... my apologies for that.

What I wanted tosay is that `throughputCapacity` and `throughputCapacityPerHaPair` have very similar roles, and with the recent improvements, `throughputCapacityPerHaPair` has become much more user-friendly. Therefore, I believe it would be simpler to remove `throughputCapacity` altogether.

go-to-k replied:
I see. If there is no use case for `throughputCapacity`, I think it's a good idea too!

go-to-k replied:
L2 does not necessarily have to address all properties and simplicity is a good thing.

badmintoncryer replied:
I've updated my code including README, integ test, jsdoc like below.

```ts
export abstract class ThroughputCapacityPerHaPair {
  /**
   * The deployment type of the throughput capacity.
   */
  public abstract readonly deploymentType: OntapDeploymentType;
  protected abstract readonly allowedCapacity: number[];
  protected constructor(public readonly capacity: number) {}

  /**
   * Check if the throughput capacity is valid for the given deployment type.
   */
  public validateCapacity(): void {
    if (!this.allowedCapacity.includes(this.capacity)) {
      throw new Error(
        `'throughputCapacity / haPairs' must be one of the following values for ${
          this.deploymentType
        }: ${this.allowedCapacity.join(', ')}. got: ${this.capacity} MB/s/HA pair`,
      );
    }
  }
}

export class SingleAz2ThroughputCapacityPerHaPair extends ThroughputCapacityPerHaPair {
  /**
   * The throughput capacity of 1536 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_1536 = new SingleAz2ThroughputCapacityPerHaPair(1536);

  /**
   * The throughput capacity of 3072 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_3072 = new SingleAz2ThroughputCapacityPerHaPair(3072);

  /**
   * The throughput capacity of 6144 MBps per HA pair.
   */
  public static readonly MB_PER_SEC_6144 = new SingleAz2ThroughputCapacityPerHaPair(6144);

  /**
   * The deployment type of the throughput capacity.
   */
  public readonly deploymentType = OntapDeploymentType.SINGLE_AZ_2;
  protected readonly allowedCapacity = [1536, 3072, 6144];
}
```

And just now, I think `ThroughputCapacityPerHaPair.validateCapacity()` may be unnecessary because all of the `throughputCapacity` is configured by `DeploymentTypeThroughputCapacityPerHaPair`. The validation of mismatch of deployment type between `deploymenType` prop and `throughputCapacity` prop is already executed at [this line](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R627). Therefore, I think it is better to remove `ThroughputCapacityPerHaPair.validateCapacity()` and [this validation](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R632).

go-to-k replied:
> And just now, I think ThroughputCapacityPerHaPair.validateCapacity() may be unnecessary because all of the throughputCapacity is configured by DeploymentTypeThroughputCapacityPerHaPair. The validation of mismatch of deployment type between deploymenType prop and throughputCapacity prop is already executed at [this line](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R627). Therefore, I think it is better to remove ThroughputCapacityPerHaPair.validateCapacity() and [this validation](https://github.com/open-constructs/aws-cdk-library/pull/42/files#diff-4411deda71b1bb070dac524d0cda1b33091f25bdc5da4d1c5fc8c86468b66d60R632).

I think so!

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+  readonly throughputCapacityPerHaPair?: number;
+
+  /**
+   * The preferred day and time to perform weekly maintenance. The first digit is the day of the week, starting at 1
+   * for Monday, then the following are hours and minutes in the UTC time zone, 24 hour clock. For example: '2:20:30'
```
```suggestion
   * The preferred day and time to perform weekly maintenance.
   *
   * The first digit is the day of the week, starting at 1
   * for Monday, then the following are hours and minutes in the UTC time zone, 24 hour clock. For example: '2:20:30'
```

go-to-k reviewed (changes requested):
I have only seen it halfway through, but I leave comments.

go-to-k commented:
@badmintoncryer 

I made a lot of comments, but glad if you could take your time and have a look at it!

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   *
+   * @default - do not set an admin password
+   */
+  readonly fsxAdminPassword?: string;
```
Can we use the `SecretValue` type?

badmintoncryer replied:
I've updated my code to use `SecretValue`.

badmintoncryer replied:
I'm not very familiar with how to use SecretValue, so could you explain it to me?

Is it correct to assume that the typical use case is passing something created with `SecretValue.secretsManager()` as an argument? In that case, my understanding is that it would no longer be possible to validate the secret itself (such as checking the length), is that correct?

go-to-k replied:
> Is it correct to assume that the typical use case is passing something created with SecretValue.secretsManager() as an argument?

Yeah, users can specify the password with `SecretValue.secretsManager` (or `SecretValue.unsafePlainText`):

- `secretsManager`

```ts
const fsxProps = {
  // ...
  fsxAdminPassword: SecretValue.secretsManager('my-secrets-arn-or-id')
}
```

- `unsafePlainText` (that is raw string, so it's unsafe)

```ts
const fsxProps = {
  // ...
  fsxAdminPassword: SecretValue.unsafePlainText('MyUnsafeAdminPassword'),
}
```

We contributor use `unsafeUnwrap()` method to extract the value in constructs.

```ts
fsxAdminPassword = props.fsxAdminPassword?.unsafeUnwrap()
```

Then the string value is printed in CFn:

```
// secretsManager
"FsxAdminPassword": "{{resolve:secretsmanager:my-secrets-arn-or-id:SecretString:::}}",

// unsafePlainText
"FsxAdminPassword": "MyUnsafeAdminPassword"
```

> In that case, my understanding is that it would no longer be possible to validate the secret itself (such as checking the length), is that correct?

Yes, we have discussed password validation, but this would no longer be necessary.

https://github.com/open-constructs/aws-cdk-library/pull/42#discussion_r1780030824

badmintoncryer replied:
Thank you for your kindly explanation! I didn't know how secret is taken in CFn.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   * The minimum value is calculated as StorageCapacity * 3 * HAPairs (3 IOPS per GB of StorageCapacity).
+   * The maximum value is calculated as 200,000 * HAPairs.
+   *
+   * @default - 3 IOPS per GB of storage capacity
```
You're right! I updated the description to `3 IOPS * GB of storage capacity * HAPairs`

go-to-k replied:
same?: https://github.com/open-constructs/aws-cdk-library/pull/42#discussion_r1781422161

go-to-k replied:
I could confirm

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   * for Monday, then the following are hours and minutes in the UTC time zone, 24 hour clock. For example: '2:20:30'
+   * is Tuesdays at 20:30.
+   *
+   * @default - no preference
```
This means automatically set! I've fixed my description.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   *
+   * @see https://docs.aws.amazon.com/fsx/latest/ONTAPGuide/managing-throughput-capacity.html
+   *
+   * @default - recommended throughput capacity based on the storage capacity
```
I was unable to locate documentation explicitly mentioning the detailed dependency between throughput capacity and storage capacity in FSx for NetApp ONTAP.
However, it's confirmed that the throughput capacity is determined based on the storage capacity.

Would a sentence like "Amazon FSx determines the throughput capacity based on the storage capacity" be more appropriate?

go-to-k replied:
> Would a sentence like "Amazon FSx determines the throughput capacity based on the storage capacity" be more appropriate?

I prefer it.

go-to-k commented:
@badmintoncryer 

You probably forgot to push the commit, so once you have done that we will wait for you. Let us know when you're done.

And don't forget the following.

https://github.com/open-constructs/aws-cdk-library/pull/42#issuecomment-2368043430

badmintoncryer commented:
@go-to-k Thank you very much for all of your comments. I plan to address all of your review but it will take some days to complete it. I'll touch you later.

go-to-k commented:
@badmintoncryer 

> @go-to-k Thank you very much for all of your comments. I plan to address all of your review but it will take some days to complete it. I'll touch you later.

Thanks, I’m looking forward to it! (And I would be happy to see CI build errors at that time.)

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+      ontapConfiguration: {
+        automaticBackupRetentionDays: ontapConfiguration.automaticBackupRetention?.toDays() ?? 0,
+        dailyAutomaticBackupStartTime: ontapConfiguration.dailyAutomaticBackupStartTime?.toTimestamp(),
+        deploymentType: ontapConfiguration.deploymentType ?? OntapDeploymentType.MULTI_AZ_2,
```
I made `ontapConfiguration.deploymentType` optional!

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+      backupId: props.backupId,
+      kmsKeyId: props.kmsKey?.keyId,
+      ontapConfiguration: {
+        automaticBackupRetentionDays: ontapConfiguration.automaticBackupRetention?.toDays() ?? 0,
```
It’s better to set up a backup as the default, so I’ve made the default value 30 days.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,520 @@
+   */
+  readonly ontapConfiguration: OntapConfiguration;
+
+  /**
+   * The subnet that the file system will be accessible from.
+   *
+   * For MULTI_AZ_1 deployment types,
+   * this subnet is for the standby file server and you have to specify a `prefferredSubnet` for the preffered file server.
+   */
+  readonly vpcSubnets: aws_ec2.ISubnet[];
```
Why I change the description from CFn's one..? I've updated to use CFn documentation.

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,518 @@
+  /**
+   * The subnet that the file system will be accessible from.
+   *
+   * For MULTI_AZ_1 deployment types,
+   * provide exactly two subnet IDs, one for the preferred file server and one for the standby file server.
+   *
+   * You specify one of these subnets as the preferred subnet using `OntapConfiguration.preferredSubnet` property for multi-AZ file system.
```
```suggestion
   * For MULTI_AZ_1 deployment types,
   * provide exactly two subnets, one for the preferred file server and one for the standby file server.
   *
   * Specify one of these subnets as the preferred subnet using `OntapConfiguration.preferredSubnet` property for multi-AZ file system.
```

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,606 @@
+  /**
+   * The throughput capacity of 128 MBps per HA pair.
+   */
+  public static readonly MB_PER_SEC_128 = ThroughputCapacityPerHaPair.MbPerSec(128);
```
Could you modify it to `mbPerSec`? (because the build failed!)

```
error JSII8002: Method and property (unless they are static readonly) names must use camelCase. Rename "@open-constructs/aws-cdk.aws_fsx.ThroughputCapacityPerHaPair.MbPerSec" to "mbPerSec"
👾 Task "build » compile" failed when executing "jsii --silence-warnings=reserved-word" (cwd: /home/runner/work/aws-cdk-library/aws-cdk-library)
```

badmintoncryer commented:
@go-to-k
I was planning to reach out once everything had settled, but I appreciate your prompt response.

I've implemented the `ThroughputCapacityPerHaPair` class and added validation logic.

However, I’m not very confident about the overall approach, so I’d like your feedback. In particular, I’m not sure about the method name `MB_PER_SEC_NUMBER`, and I think it could be improved.

go-to-k commented:
> In particular, I’m not sure about the method name MB_PER_SEC_NUMBER, and I think it could be improved.

- MB_PER_SEC_128
- PER_HA_PAIR_128_MB
- VALUE_128_MB_PER_SEC
- SIZE_128_MB_PER_SEC
- CAPACITY_128_MB_PER_SEC
- THROUGHPUT_128_MB_PER_SEC
- THROUGHPUT_CAPACITY_128_MB_PER_SEC
- etc...

Hmmm... That's difficult... I prefer the format of `XX_128_MB_PER_SEC` or `XX_128_MB`, but `MB_PER_SEC_128` is also good...

go-to-k commented:
Please also change the README, JSDOC, and tests later (if we need to change anything to match your implementation).

go-to-k commented:
Let's go with MB_PER_SEC_NUMBER for now!

go-to-k commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,649 @@
+  /**
+   * The throughput capacity per HA pair for the file system.
+   *
+   * For SINGLE_AZ_1 and MULTI_AZ_1 file systems, valid values are 128, 256, 512, 1024, 2048, or 4096 MBps.
+   * For SINGLE_AZ_2, valid values are 1536, 3072, or 6144 MBps.
+   * For MULTI_AZ_2, valid values are 384, 768, 1536, 3072, or 6144 MBps.
+   *
```
It is forgotten to maintain it in the future when the number of types increases and users can know about it in the enum-like class. In other words, this is good to remove it.

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
You're right!

badmintoncryer commented on src/aws-fsx/ontap-file-system.ts:
```diff
@@ -0,0 +1,649 @@
+   * The deployment type of the throughput capacity.
+   */
+  public abstract readonly deploymentType: OntapDeploymentType;
+  protected abstract readonly allowedCapacity: number[];
```
What is the meaning?

go-to-k replied:
That means the allowedCapacity can be removed.

badmintoncryer replied:
You're absolutely right! I've removed it.

go-to-k reviewed (approved):
Could confirmed all changes. Thanks for this PR. I approve and merge.

badmintoncryer commented:
@go-to-k Thank you very much for your review!
